core/include/drivers/nvmem.h

```diff
@@ -0,0 +1,188 @@
+/**
+ * nvmem_register_provider() - Register a nvmem controller
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of nvmem cell consumer
+ * @get_dt_pinctrl: Callback to match the devicetree nvmem reference with
+ * nvmem_cell
+ * @data: Data which will be passed to the get_dt_nvmem callback
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result nvmem_register_provider(const void *fdt,
+						 int nodeoffset,
+						 nvmem_dt_get_func get_dt_nvmem,
+						 void *data)
+{
+	return dt_driver_register_provider(fdt, nodeoffset,
+					   (get_of_device_func)get_dt_nvmem,
+					   data, DT_DRIVER_NVMEM);
+}
+
+/**
+ * nvmem_get_cell_by_name() - Obtain a nvmem cell from its name in the DT node
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of nvmem cell consumer
+ * @name: name of the nvmem cell to obtain from the device tree
+ * @cell: Pointer filled with the retrieved cell, must be freed after use
+   using nvmem_free_cell()
+ * Return a TEE_Result compliant value
+ */
+TEE_Result nvmem_get_cell_by_name(const void *fdt, int nodeoffset,
+				  const char *name, struct nvmem_cell **cell);
+
+/**
+ * nvmem_get_cell_by_index() - Obtain a nvmem cell from property nvmem-cells
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of nvmem cell consumer
+ * @index: Index of the nvmem cell to obtain from device-tree
+ * @cell: Pointer filled with the retrieved cell, must be freed after use
+ * using nvmem_free_cell()
+ * Return a TEE_Result compliant value
+ */
+static inline TEE_Result nvmem_get_cell_by_index(const void *fdt,
+						 int nodeoffset,
+						 unsigned int index,
+						 struct nvmem_cell **cell)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	*cell = dt_driver_device_from_node_idx_prop("nvmem-cells", fdt,
+						    nodeoffset, index,
+						    DT_DRIVER_NVMEM, &res);
+	return res;
+}
+
+/**
+ * nvmem_cell_parse_dt() - Parse device-tree information to fill a nvmem cell
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the nvmem cell controller
+ * @cell: Pointer to cell that will be filled
+ */
+TEE_Result nvmem_cell_parse_dt(const void *fdt, int nodeoffset,
+			       struct nvmem_cell *cell);
+
+/**
+ * nvmem_put_cell() - Free resource allocated from nvmem_get_cell_by_*()
+ *
+ * @cell: Cell to be freed
+ */
+static inline void nvmem_put_cell(struct nvmem_cell *cell)
+{
+	if (cell->ops->free_cell)
+		cell->ops->free_cell(cell);
+}
+
+/*
+ * nvmem_cell_alloc_read() - Allocate and read data from a nvmem cell
+ * @cell: Cell to read from nvmem
+ * @data: Output allocated buffer where nvmem cell data are stored upon success,
+ * shall be released with free().
+ * @len: Output byte size of the nvmem cell data read
+ */
+static inline TEE_Result nvmem_cell_alloc_read(struct nvmem_cell *cell,
+					       uint8_t **data, size_t *len)
+{
+	if (!cell->ops->read_cell)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	return cell->ops->read_cell(cell, data, len);
```

Comment:
Caller of `nvmem_cell_alloc_read()` knows the cell byte size (`struct nvmem_cell::len`) so caller is fully able to provide a well sized buffer to load the cell byte content into. This removes need for allocation/free from the framework. 
If i'm right, could you updat the interface accordingly?
Rename `nvmem_cell_alloc_read()` to original `nvmem_cell_read()` and I would keep the `len` argument for sanity test by backend handler function.